src/s3iomain.cpp

```diff
@@ -0,0 +1,172 @@
+
+    char buffer[WINDOW_SIZE]{0};
+    auto readBytes = fread(buffer, sizeof(char), sizeof(buffer), filePointer);
+    if (readBytes > 0 && '\n' == buffer[readBytes - 1])
+        buffer[readBytes - 1] = '\0';
+    fclose(filePointer);
+    return {true, string(buffer)};
+}
+
+/**
+ * Get the pid of this applications parent process.
+ *
+ * @see https://gist.github.com/fclairamb/a16a4237c46440bdb172
+ */
+tuple<bool, pid_t> getProcessParentID(const pid_t pid) {
+    auto[filePointer, file] = openProcOrUserFile(pid, "stat");
+    if (!filePointer) return {false, 0};
+
+    char buffer[WINDOW_SIZE]{0};
+    auto readBytes = fread(buffer, sizeof(char), sizeof(buffer), filePointer);
+
+    if (readBytes > 0) {
+        // See: http://man7.org/linux/man-pages/man5/proc.5.html section /proc/[pid]/stat
+        strtok(buffer, " ");           // (1) pid  %d
+        strtok(nullptr, " ");       // (2) comm  %s
+        strtok(nullptr, " ");       // (3) state  %c
+        char *s_ppid = strtok(nullptr, " "); // (4) ppid  %d
+
+        return {true, atoi(s_ppid)};
+    }
+    fclose(filePointer);
+    return {false, 0};
+}
+
+/**
+ * Returns the parent path of the this binary.
+ */
+tuple<bool, path> getParentPath() {
+    int myPid = getpid();
+    auto[processParentIDValid, processParentID] = getProcessParentID(myPid);
+    if(!processParentIDValid) {
+        std::cerr << "Could not get process id for parent process.\n";
+        return {false, ""};
+    }
+
+    auto[processNameValid, processName] = getProcessName(processParentID);
+    if(!processNameValid) {
+        std::cerr << "Could not get the name of the process.\n";
+        return {false, ""};
+    }
+
+    return {true, processName};
+}
+
+/**
+ * The main function will 6 parameters (+1 for the application path in parameter 0).
+ * The application is not meant for standalone usage and does not feature help messages.
+ * @param argv  The argument list for the program. The necessary values are:
+ *              [0] - Application path / set by the system (actually not necessary)
+ *              [1] - The fifo or file to which the GetObject request will write.
+ *              [2] - The S3 object like in s3://bucket/object - It is the same as for the fastqindex binary.
+ *              [3] - The S3 configuration sections or "default"
+ *              [4] - The S3 configuration file or (IMPORTANT) " " <- See the whitespace inside double ticks?
+ *              [5] - The S3 credentials file or (ALSO HERE) " " <- Again the whitespace surrounded by " " double ticks!
+ *              [6] - The starting position.
+ * @return Some value, most likely an error value as we are not able to properly abort S3 GetObject
+ */
+int main(int argc, char **argv) {
+    auto [parentPathRetrieved, parentPath] = getParentPath();
+    if(!parentPathRetrieved) {
+        std::cerr << "Could not get the parent processes path.\n";
+        return -1;
+    }
+    if (parentPath != "fastqindex") {
```

Comment:
Actually, I think the name of also this binary should be camel-case. All lower-case is also less readable in Bash (although Bash itself may be the larger burden).